backend/corpora/common/entities/project.py

```diff
@@ -0,0 +1,129 @@
+
+    @classmethod
+    def _query(cls, key: typing.Tuple[str, str]) -> typing.List[Base]:
+        """
+        Given a key, queries the database for a project and its relevant entities.
+
+        According to the Entity.get interface, the return value of this function is fed as the input to Project._load.
+
+        :param key: Composite primary key tuple of the form (project.id, project.status)
+        :return: list of SQLAlchemy query results
+        """
+        statuses = [status.name for status in ProjectStatus]
+
+        if key[1] not in statuses:
+            raise CorporaException(f"Invalid status {key[1]}. Status must be one of {statuses}.")
+
+        result = cls.db.query(
+            table_args=[DbProject, DbProjectLink, DbProjectDataset, DbDatasetContributor, DbContributor],
+            filter_args=[
+                DbProject.id == key[0],
+                DbProject.status == key[1],
+                DbProject.id == DbProjectLink.project_id,
+                DbProject.status == DbProjectLink.project_status,
+                DbProject.id == DbProjectDataset.project_id,
+                DbProject.status == DbProjectDataset.project_status,
+                DbProjectDataset.dataset_id == DbDatasetContributor.dataset_id,
+                DbContributor.id == DbDatasetContributor.contributor_id,
+            ],
+        )
+
+        return result
+
+    @classmethod
+    def _load(cls, query_results: typing.List[Base]) -> typing.Union["Project", None]:
+        """
+        Parses query result rows produced by Project._query into an instantiated Project object.
+        The output of this function is the return value of Entity.get.
+
+        SQLAlchemy query results are stored in a list in which each item contains Table objects
+        (Db* objects from corpora_orm.py) returned by the query.
+
+        Example query results access:
+        row = query_results[0]
+        project_id = row.DbProject.id
+
+        :param query_results: list of query result rows
+        :return: Project
+        """
+        if not query_results:
+            return None
+
+        # de-dupe peripheral entities from query results
+        dataset_ids = set()
+        links = {}
+        contributors = {}
+        for result in query_results:
+            dataset_ids.add(result.DbProjectDataset.dataset_id)
+
+            links[result.DbProjectLink.id] = {
+                "id": result.DbProjectLink.id,
+                "type": result.DbProjectLink.link_type,
+                "url": result.DbProjectLink.link_url,
+            }
+
+            contributors[result.DbContributor.id] = {
+                "id": result.DbContributor.id,
+                "name": result.DbContributor.name,
+                "institution": result.DbContributor.institution,
+                "email": result.DbContributor.email,
+            }
+
+        # build dict of Project parameters
+        project_params = query_results[0].DbProject.__dict__.copy()
```

Comment:
You're probably going to want to deepcopy here instead of shallow copy.

Reply:
Done